common/include/camera/ICameraControl.hpp

```diff
@@ -0,0 +1,312 @@
+#include "glm/glm/gtc/quaternion.hpp"
+#include "nbl/builtin/hlsl/matrix_utils/transformation_matrix_utils.hlsl"
+
+// TODO: DIFFERENT NAMESPACE
+namespace nbl::hlsl 
+{
+
+template<ProjectionMatrix T = float64_t4x4>
+class ICameraController : virtual public core::IReferenceCounted
+{
+public:
+    using projection_t = typename IProjection<typename T>;
+
+    enum VirtualEventType : uint8_t
+    {
+        // Strafe forward
+        MoveForward = 0,
+
+        // Strafe backward
+        MoveBackward,
+
+        // Strafe left
+        MoveLeft,
+
+        // Strafe right
+        MoveRight,
+
+        // Strafe up
+        MoveUp,
+
+        // Strafe down
+        MoveDown,
+
+        // Tilt the camera upward (pitch)
+        TiltUp,
+
+        // Tilt the camera downward (pitch)
+        TiltDown,
+
+        // Rotate the camera left around the vertical axis (yaw)
+        PanLeft,
+
+        // Rotate the camera right around the vertical axis (yaw)
+        PanRight,
+
+        // Roll the camera counterclockwise around the forward axis (roll)
+        RollLeft,
+
+        // Roll the camera clockwise around the forward axis (roll)
+        RollRight,
+
+        // Reset the camera to the default state
+        Reset,
+
+        EventsCount
+    };
+
+    //! Virtual event representing a manipulation
+    struct CVirtualEvent
+    {
+        using manipulation_encode_t = float64_t;
+
+        VirtualEventType type;
+        manipulation_encode_t value;
+    };
+
+    class CGimbal : virtual public core::IReferenceCounted
+    {
+    public:
+        CGimbal(const float32_t3& position, glm::quat orientation = glm::quat(1.0f, 0.0f, 0.0f, 0.0f))
+            : m_position(position), m_orientation(orientation) { updateOrthonormalMatrix(); }
+
+        inline void setPosition(const float32_t3& position)
+        {
+            m_position = position;
+        }
+
+        inline void setOrientation(const glm::quat& orientation)
+        {
+            m_orientation = glm::normalize(orientation);
+            updateOrthonormalMatrix();
+        }
+
+        inline void rotate(const float32_t3& axis, float dRadians)
+        {
+            glm::quat dRotation = glm::angleAxis(dRadians, axis);
+            m_orientation = glm::normalize(dRotation * m_orientation);
+            updateOrthonormalMatrix();
+        }
+
+        inline void move(float32_t3 delta)
+        {
+            m_position += delta;
+        }
+
+        inline void reset()
+        {
+            // TODO
+        }
+
+        // Position of gimbal
+        inline const float32_t3& getPosition() const { return m_position; }
+
+        // Orientation of gimbal
+        inline const glm::quat& getOrientation() const { return m_orientation; }
+
+        // Orthonormal [getXAxis(), getYAxis(), getZAxis()] matrix
+        inline const float32_t3x3& getOrthonornalMatrix() const { return m_orthonormal; }
```

Comment:
to avoid semantical confusion (as to what your postion means) I'd also add a `getMatrix()` method which would return a `float64_t3x4` (a full TRS).

P.s. if you like you can even make it an `operator float64_t3x4`

Reply:
resolved